File: go/protein-translation/protein_translation.go
package protein

import "errors"

// ErrStop is an error for stop codons.
var ErrStop = errors.New("stop codon")

// ErrInvalidBase is an error for invalid strings
var ErrInvalidBase = errors.New("invalid base")

var m = map[string]string{
	"AUG": "Methionine",
	"UUU": "Phenylalanine", "UUC": "Phenylalanine",
	"UUA": "Leucine", "UUG": "Leucine",
	"UCU": "Serine", "UCC": "Serine", "UCA": "Serine", "UCG": "Serine",
	"UAU": "Tyrosine", "UAC": "Tyrosine",
	"UGU": "Cysteine", "UGC": "Cysteine",
	"UGG": "Tryptophan",
	"UAA": "STOP", "UAG": "STOP", "UGA": "STOP",
}

// FromCodon returns a nucleotide sequence
func FromCodon(codon string) (string, error) {
	if _, ok := m[codon]; ok {
		if m[codon] == "STOP" {
			return "", ErrStop
		}
		return m[codon], nil
	}
	return "", ErrInvalidBase
}

// FromRNA returns an array of nucleotide sequences
func FromRNA(protein string) ([]string, error) {
	res := []string{}
	for i := 0; i < len(protein)-2; i += 3 {
		test := string(protein[i : i+3])
		testRes, err := FromCodon(test)
		if err == nil {
			res = append(res, testRes)
		} else {
			if err == ErrStop {
				return res, nil
			}
			return res, ErrInvalidBase
		}
	}
	return res, nil
}

File: swift/protein-translation/Sources/ProteinTranslation/ProteinTranslation.swift
enum TranslationError: Error {
    case invalidArgument
}

class ProteinTranslation {
    static var proteinMap = [
        "AUG": "Methionine",
        "UUU": "Phenylalanine", "UUC": "Phenylalanine",
        "UUA": "Leucine", "UUG": "Leucine",
        "UCU": "Serine", "UCC": "Serine", "UCA": "Serine", "UCG": "Serine",
        "UAU": "Tyrosine", "UAC": "Tyrosine",
        "UGU": "Cysteine",  "UGC": "Cysteine",
        "UGG": "Tryptophan",
        "UAA": "STOP", "UAG": "STOP", "UGA": "STOP"
    ]
    
    static func translationOfCodon(_ codon: String) throws -> String {
        guard let res = self.proteinMap[codon] else {
            throw TranslationError.invalidArgument
        }
        
        return res
    }
    
    static func translationOfRNA(_ rna: String) throws -> [String] {
        var res = [String]()
        var i = 0

        while i <= rna.count - 3 {
            let start = rna.index(rna.startIndex, offsetBy: i)
            let end = rna.index(rna.startIndex, offsetBy: i + 2)
            let codon = String(rna[start...end])

            if (self.proteinMap[codon] == "STOP") {
                return res
            }

            res.append(try translationOfCodon(codon))
            i += 3
        }
        return res
    }
}
